cumulusci/tasks/salesforce/package_upload.py

```diff
@@ -0,0 +1,222 @@
+        if "description" in self.options:
+            package_info["Description"] = self.options["description"]
+        if "password" in self.options:
+            package_info["Password"] = self.options["password"]
+        if "post_install_url" in self.options:
+            package_info["PostInstallUrl"] = self.options["post_install_url"]
+        if "release_notes_url" in self.options:
+            package_info["ReleaseNotesUrl"] = self.options["release_notes_url"]
+
+    def _make_package_upload_request(self, package_info, package_id):
+        """Creates a PackageUploadRequest in self.upload"""
+        PackageUploadRequest = self._get_tooling_object("PackageUploadRequest")
+
+        self._upload_start_time = time.time()
+        self.upload = PackageUploadRequest.create(package_info)
+
+        self.upload_id = self.upload["id"]
+        self.logger.info(
+            f"Created PackageUploadRequest {self.upload_id} for Package {package_id}"
+        )
+        self._poll()
+
+    def _log_package_upload_errors(self):
+        self.logger.error("Package upload failed with the following errors")
+        error = {"message": ""}
+        apex_test_failures = False
+        for error in self.upload["Errors"]["errors"]:
+            self.logger.error("  {}".format(error["message"]))
+            if error["message"] == "ApexTestFailure":
+                apex_test_failures = True
+
+        if apex_test_failures:
+            self._log_apex_test_failures()
+
+        exception = self._get_exception_type(error)
+        raise exception("Package upload failed")
+
+    def _log_apex_test_failures(self):
+        self.logger.error("Failed Apex Tests:")
+        soql_query = self._get_failed_tests_soql_query()
+        results = self.tooling.query(soql_query)
+
+        table_header_row = ["Class", "Method", "Stacktrace"]
+        table_data = [table_header_row]
+        for test in results["records"]:
+            table_data.append(
+                [test["ApexClass"]["Name"], test["MethodName"], test["StackTrace"]]
+            )
+
+        table_title = "Failed Apex Tests"
+        table = CliTable(table_data, table_title, wrap_cols=["Stacktrace"])
+        table.echo()
+
+    def _get_failed_tests_soql_query(self):
+        package_upload_datetime = self._get_package_upload_iso_timestamp()
+        return (
+            "SELECT ApexClass.Name, "
+            "MethodName, "
+            "Message, "
+            "StackTrace "
+            "FROM ApexTestResult "
+            "WHERE Outcome='Fail' "
+            f"AND TestTimestamp > {package_upload_datetime}Z"
+        )
+
+    def _get_package_upload_iso_timestamp(self):
+        """Returns a datetime of approximately when package upload began
+        This assumes a short time between the call to
+        _make_package_upload_request() and this function"""
+        test_start_datetime = datetime.utcnow() - timedelta(
+            seconds=(time.time() - self._upload_start_time)
+        )
```

Comment:
I still don't understand the reason for the date arithmetic. Why not just set `self._upload_start_time = datetime.utcnow()` up above, and use self._upload_start_time.isoformat() for the query?

Reply:
Ahhh. Yes...was definitely overcomplicating this.